src/OVAL/probes/independent/environmentvariable_probe.c

```diff
@@ -84,88 +84,77 @@ static int read_environment(SEXP_t *un_ent, probe_ctx *ctx)
 
 #else // EXTERNAL_PROBE_COLLECT
 
-static int extract_matching_env_keys(SEXP_t* un_ent, probe_ctx* ctx, oval_external_probe_value_map_t* env_vals) {
-    int err = 0;
-    SEXP_t *env_name, *env_value, *item;
-
-    if (env_vals == NULL)
-        return 0;
-
-    const char* name;
-    oval_external_probe_value_t* val;
-    OVAL_EXTERNAL_PROBE_VALUE_MAP_FOREACH(env_vals, name, val, {
-        if (oval_external_probe_value_get_datatype(val) != OVAL_DATATYPE_STRING) {
-            err = PROBE_EINVAL;
-            break;
-        }
-
-        env_name = SEXP_string_newf("%s", name);
-        env_value = SEXP_string_newf("%s", oval_external_probe_value_get_string(val));
-        if (probe_entobj_cmp(un_ent, env_name) == OVAL_RESULT_TRUE) {
-            dI("EXTPROBE: environmentvariable: matched env_name=%s, value=%s", name, val);
-            item = probe_item_create(OVAL_INDEPENDENT_ENVIRONMENT_VARIABLE, NULL, "name", OVAL_DATATYPE_SEXP, env_name, "value", OVAL_DATATYPE_SEXP,
-                                     env_value, NULL);
-            probe_item_collect(ctx, item);
-        }
-        SEXP_free(env_name);
-        SEXP_free(env_value);
-    })
-
-    return err;
-}
-
-static int read_environment(SEXP_t *un_ent, probe_ctx *ctx)
-{
-    oval_external_probe_eval_funcs_t* eval = probe_get_external_probe_eval(ctx);
-    if (eval == NULL || eval->environmentvariable_probe == NULL)
-		return PROBE_EOPNOTSUPP;
-
-	SEXP_t *probe_in = probe_ctx_getobject(ctx);
-    SEXP_t* oid = NULL;
-    oval_external_probe_result_t* res = NULL;
-	int err = PROBE_ENOVAL;
-
-    oid = probe_obj_getattrval(probe_in, "id");
-	if (oid == NULL)
-		goto cleanup;
-
-	char *id = SEXP_string_cstr(oid);
-	res = eval->environmentvariable_probe(eval->probe_ctx, id);
-	free(id);
-
-	if (res == NULL) {
-		goto cleanup;
-	}
-
-	err = oval_external_probe_result_get_status(res);
-	if (err != 0)
-		goto cleanup;
-
-	err = extract_matching_env_keys(un_ent, ctx, oval_external_probe_result_get_fields(res));
-
-cleanup:
-	oval_external_probe_result_free(res);
-	SEXP_free(oid);
-
-	return err;
+static int read_environment(SEXP_t *un_ent, probe_ctx *ctx) {
+    int ret;
+    char *str_id = NULL;
+    SEXP_t *in, *id = NULL;
+    oval_syschar_status_t status;
+    oval_external_probe_eval_funcs_t *eval;
+    oval_external_probe_item_t *ext_query = NULL;
+    oval_external_probe_result_t *ext_res = NULL;
+    oval_external_probe_item_list_t *ext_items;
+
+    __attribute__nonnull__(ctx);
+
+    eval = probe_get_external_probe_eval(ctx);
+    if(eval == NULL || eval->environment_variable_probe == NULL) {
+        ret = PROBE_EOPNOTSUPP;
+        goto fail;
+    }
+    in = probe_ctx_getobject(ctx);
+    id = probe_obj_getattrval(in, "id");
+    if(id == NULL) {
+        ret = PROBE_ENOVAL;
+        goto fail;
+    }
```

Comment:
This part is missing from the new implementation. It's mandatory to collect only the env vars matching the name provided by `un_ent`.

Reply:
I've left it to the external probe to interpret the query and return only the matching env variables. I'm not sure what approach is better, so, if you think it's better to implement/readd that logic here, I'll do it.

Reply:
I've backed up the new commits on a new branch and reverted the changes. The specialised external probes will be removed in a future PR.